File: scripts/scoring.py
from statistics import fmean, pstdev

# Weight given to a targeted opponent (u>1) beyond its actual severity. Fixed at
# the severity of a 4.5 "slight disadvantage" (sev = (5-4.5)^2 = 0.25) so the
# target slider behaves consistently regardless of how bad the main's worst
# matchup is. Only applies when u>1; u=1 (normal) never injects.
TARGET_INJECT = 0.25


def expand_months(args):
    """['202601-202605'] or ['202604', '202605'] -> explicit month list."""
    out = []
    for a in args:
        if '-' in a:
            lo, hi = a.split('-', 1)
            y, mo = int(lo[:4]), int(lo[4:])
            while f'{y}{mo:02d}' <= hi:
                out.append(f'{y}{mo:02d}')
                mo += 1
                if mo == 13:
                    y, mo = y + 1, 1
        else:
            out.append(a)
    return out


def month_weights(months, profile, patch_month):
    if profile == 'all':
        return {m: 1.0 for m in months}
    if profile == 'current':
        w = {m: 0.0 if m < patch_month else 0.5 if m == patch_month else 1.0
             for m in months}
        return w if any(w.values()) else {m: 1.0 for m in months}
    raise ValueError(f'unknown profile: {profile}')


def parse_weights(spec):
    """'202601=0,202603=0.5' -> {'202601': 0.0, '202603': 0.5}.
    The {month: weight} dict is the recalculation input shared by the CLIs
    and any future interactive frontend."""
    return {m: float(w) for part in spec.split(',')
            for m, w in [part.split('=')]}


def wavg(scores, weights):
    """Weighted average of scores over keys with nonzero weight; None if empty."""
    pairs = [(scores[k], weights[k]) for k in scores if weights.get(k)]
    if not pairs:
        return None
    return sum(s * w for s, w in pairs) / sum(w for _, w in pairs)


def usage_weights(rates):
    """Per-opponent usage weight from {char: play_rate}: sqrt(rate / mean_rate),
    so an average-popularity opponent = 1.0, popular > 1, rare < 1 (dampened by
    the sqrt so a 9%-played character doesn't dwarf a 1% one 9x). Empty -> {}."""
    if not rates:
        return {}
    mean = fmean(rates.values())
    return {c: (r / mean) ** 0.5 for c, r in rates.items()} if mean else {}


def _weakness_weights(main_row, opp_weights=None, usage=None):
    """w(O) per opponent for the coverage family:
        w(O) = u(O) * usage(O) * sev(O) + max(0, u(O) - 1) * TARGET_INJECT
        sev(O) = max(0, 5 - main_vs_O) ** 2          # weakness severity

    u(O) is the per-opponent user weight (default 1.0): u=1 gives the plain
    weakness-weighted scheme, u=0 drops the opponent, u>1 targets it (counts even
    when it is not a current weakness, via TARGET_INJECT). usage(O) is an optional
    popularity multiplier (default 1.0) that down-weights rarely-played opponents.
    Returns only opponents with nonzero weight."""
    out = {}
    for opp, ms in main_row.items():
        sev = max(0.0, 5.0 - ms) ** 2
        if usage:
            sev *= usage.get(opp, 1.0)
        u = 1.0 if opp_weights is None else opp_weights.get(opp, 1.0)
        w = u * sev + max(0.0, u - 1.0) * TARGET_INJECT
        if w:
            out[opp] = w
    return out


def strength(row):
    """Overall mean matchup score of a character's combined row (tier proxy).
    Empty row -> neutral 5.0."""
    return fmean(row.values()) if row else 5.0


def pair_coverage(main_row, sub_row, thresh=5.0):
    """How well a main+partner ROSTER covers the cast: for each opponent you field
    whichever of the two has the better matchup. 'covered' = opponents the pair wins
    (max score >= thresh, default 5.0 = even-or-better); 'patched' = main's losing
    matchups (< thresh) the partner flips to a win; 'losses' = main's losing matchups;
    'total' = opponents in the main's row. Opponents missing from sub_row fall back to
    the main alone (the partner can't help there). Keep in sync with web/scoring.js."""
    total = len(main_row)
    losses = covered = patched = 0
    for opp, ms in main_row.items():
        if ms < thresh:
            losses += 1
            if sub_row.get(opp, 0.0) >= thresh:
                patched += 1
        if max(ms, sub_row.get(opp, ms)) >= thresh:
            covered += 1
    return {'total': total, 'losses': losses, 'patched': patched, 'covered': covered}


def polarization(row):
    """Spread of a character's combined matchup row = population std of its scores.
    High = feast-or-famine (hard counters and free wins); low = even across the cast.
    Empty or single-opponent rows -> 0.0. Keep in sync with web/scoring.js."""
    v = list(row.values())
    return pstdev(v) if len(v) > 1 else 0.0


def reliability(nmo, nranks, spread):
    """How much agreeing data backs a single combined matchup number.

    The Buckler source does not publish sample sizes, so confidence is derived
    from what we do observe: contributing months (data depth), rank-tier coverage
    (breadth), and cross-rank spread (do the four skill brackets agree?). Returns
    {'level': 'high'|'med'|'low', 'score': 0..1}. A fully-sampled cell never falls
    to 'low' on rank disagreement alone; 'low' is reserved for genuinely thin data.

    Keep in sync with reliability() in web/scoring.js (parity-tested)."""
    depth = min(nmo, 3) / 3 * 0.4 + min(nranks, 4) / 4 * 0.3   # 0..0.7 coverage
    agree = max(0.0, 1.0 - spread / 0.6) * 0.3                 # 0..0.3 agreement
    score = depth + agree                                      # 0..1
    level = 'high' if score >= 0.85 else 'low' if score < 0.6 else 'med'
    return {'level': level, 'score': score}


def coverage(main_row, sub_row, opp_weights=None, usage=None):
    """Sub coverage of main's weaknesses: weakness-weighted ABSOLUTE edge.

    COVER = sum(w(O) * (sub_vs_O - 5)) / sum(w(O)), with w(O) from
    _weakness_weights(). Opponents missing from sub_row count as neutral (5.0)."""
    w = _weakness_weights(main_row, opp_weights, usage)
    num = sum(wt * (sub_row.get(opp, 5.0) - 5.0) for opp, wt in w.items())
    den = sum(w.values())
    return num / den if den else 0.0


def specialization(main_row, sub_row, opp_weights=None, usage=None):
    """Sub coverage measured RELATIVE TO the sub's own average matchup, so a
    globally-strong character (wins against everyone) nets ~0 and only genuine
    counters to the main's weaknesses score high. Removes the raw-strength bias
    in COVER.

    SPEC = sum(w(O) * (sub_vs_O - sub_mean)) / sum(w(O)), same w(O) as coverage.
    Opponents missing from sub_row count as neutral relative to the sub (sub_mean)."""
    sub_mean = strength(sub_row)
    w = _weakness_weights(main_row, opp_weights, usage)
    num = sum(wt * (sub_row.get(opp, sub_mean) - sub_mean) for opp, wt in w.items())
    den = sum(w.values())
    return num / den if den else 0.0


def correlation(a, b):
    """Pearson correlation over shared opponents. Negative = complementary."""
    keys = sorted(set(a) & set(b))
    if not keys:
        return 0.0
    xs, ys = [a[k] for k in keys], [b[k] for k in keys]
    mx, my = fmean(xs), fmean(ys)
    sxy = sum((x - mx) * (y - my) for x, y in zip(xs, ys))
    sxx = sum((x - mx) ** 2 for x in xs)
    syy = sum((y - my) ** 2 for y in ys)
    return sxy / (sxx * syy) ** 0.5 if sxx and syy else 0.0


def shared_weaknesses(a, b, thresh=4.9):
    return sorted(k for k in set(a) & set(b) if a[k] < thresh and b[k] < thresh)

File: web/scoring.js
/* Port of scripts/scoring.py + analyze.py/recommend.py aggregation.
   UI-free so node can require() it — tests/test_js_parity.py asserts these
   functions match the Python originals. Keep formulas in sync with METHOD.md. */

const PATCH_MONTH = '202603'; // major all-character balance patch, 2026-03-17
const DEFAULT_TIER_WEIGHTS = { 40: 1, 41: 2, 42: 3 };
const RANK_NAMES = { 40: 'HighM', 41: 'GrandM', 42: 'UltM' };
const TARGET_INJECT = 0.25;   // weight a targeted (u>1) non-weakness as a 4.5 matchup; keep in sync with scoring.py

function monthWeights(months, profile, patchMonth) {
  if (profile === 'all') return Object.fromEntries(months.map(m => [m, 1.0]));
  // 'current': pre-patch 0, patch month 0.5, post-patch 1; fallback to equal
  const w = Object.fromEntries(months.map(m =>
    [m, m < patchMonth ? 0.0 : m === patchMonth ? 0.5 : 1.0]));
  return Object.values(w).some(v => v) ? w
    : Object.fromEntries(months.map(m => [m, 1.0]));
}

function wavg(scores, weights) {
  let num = 0, den = 0;
  for (const k of Object.keys(scores)) {
    const w = weights[k];
    if (w) { num += scores[k] * w; den += w; }
  }
  return den ? num / den : null;
}

// w(O) = u(O)·sev(O) + max(0, u(O)−1)·TARGET_INJECT,  sev = max(0, 5−main)².
// u defaults to 1 (→ plain weakness-weighted scheme), u=0 drops O, u>1 targets it.
function weaknessWeights(mainRow, oppWeights) {
  const w = {};
  for (const [o, ms] of Object.entries(mainRow)) {
    const sev = Math.max(0, 5.0 - ms) ** 2;
    const u = oppWeights ? (oppWeights[o] ?? 1) : 1;
    const wt = u * sev + Math.max(0, u - 1) * TARGET_INJECT;
    if (wt) w[o] = wt;
  }
  return w;
}

// overall mean matchup score of a row (tier proxy); empty -> neutral 5.0
function strength(row) {
  const v = Object.values(row);
  return v.length ? v.reduce((s, x) => s + x, 0) / v.length : 5.0;
}

function coverage(mainRow, subRow, oppWeights) {
  // weakness-weighted ABSOLUTE edge. Missing subRow entries count as neutral (5.0).
  const w = weaknessWeights(mainRow, oppWeights);
  let num = 0, den = 0;
  for (const [opp, wt] of Object.entries(w)) {
    num += wt * ((subRow[opp] ?? 5.0) - 5.0);
    den += wt;
  }
  return den ? num / den : 0.0;
}

function specialization(mainRow, subRow, oppWeights) {
  // coverage measured relative to the sub's own average, so a globally-strong
  // character nets ~0 and only genuine counters score high. Missing subRow
  // entries count as neutral relative to the sub (subMean).
  const subMean = strength(subRow);
  const w = weaknessWeights(mainRow, oppWeights);
  let num = 0, den = 0;
  for (const [opp, wt] of Object.entries(w)) {
    num += wt * ((subRow[opp] ?? subMean) - subMean);
    den += wt;
  }
  return den ? num / den : 0.0;
}

function correlation(a, b) {
  const keys = Object.keys(a).filter(k => k in b).sort();
  if (!keys.length) return 0.0;
  const xs = keys.map(k => a[k]), ys = keys.map(k => b[k]);
  const mx = xs.reduce((s, v) => s + v, 0) / xs.length;
  const my = ys.reduce((s, v) => s + v, 0) / ys.length;
  let sxy = 0, sxx = 0, syy = 0;
  for (let i = 0; i < keys.length; i++) {
    sxy += (xs[i] - mx) * (ys[i] - my);
    sxx += (xs[i] - mx) ** 2;
    syy += (ys[i] - my) ** 2;
  }
  return sxx && syy ? sxy / Math.sqrt(sxx * syy) : 0.0;
}

function sharedWeaknesses(a, b, thresh = 4.9) {
  return Object.keys(a).filter(k => k in b && a[k] < thresh && b[k] < thresh).sort();
}

/* CSV rows -> idx[char][opp][rank][month] = score */
function buildIndex(csvText) {
  const idx = {};
  const lines = csvText.trim().split('\n');
  for (let i = 1; i < lines.length; i++) {
    const [month, rank, char, opp, score] = lines[i].split(',');
    (((idx[char] ??= {})[opp] ??= {})[rank] ??= {})[month] = parseFloat(score);
  }
  return idx;
}

function availableMonths(csvText) {
  const months = new Set();
  const lines = csvText.trim().split('\n');
  for (let i = 1; i < lines.length; i++) months.add(lines[i].split(',', 1)[0]);
  return [...months].sort();
}

/* {opp: tier-combined score} — mirrors analyze.combined_row */
function combinedRow(idx, char, mw, exclude, tierWeights, ranks = null) {
  ranks ??= Object.keys(tierWeights);
  const out = {};
  for (const [opp, byrank] of Object.entries(idx[char] ?? {})) {
    if (exclude.has(opp)) continue;
    let num = 0, den = 0;
    for (const r of ranks) {
      const v = wavg(byrank[r] ?? {}, mw);
      if (v !== null) { num += v * tierWeights[r]; den += tierWeights[r]; }
    }
    if (den) out[opp] = num / den;
  }
  return out;
}

/* mirrors analyze.char_table; rows sorted by comb ascending */
function charTable(idx, char, mw, exclude, tierWeights, patchMonth) {
  const rows = [];
  for (const [opp, byrank] of Object.entries(idx[char] ?? {})) {
    if (exclude.has(opp)) continue;
    const tier = {};
    let num = 0, den = 0;
    for (const r of Object.keys(tierWeights)) {
      tier[r] = wavg(byrank[r] ?? {}, mw);
      if (tier[r] !== null) { num += tier[r] * tierWeights[r]; den += tierWeights[r]; }
    }
    if (!den) continue;
    const present = Object.values(tier).filter(v => v !== null);
    const g = byrank['41'] ?? {};
    const pre = [], post = [];
    for (const [m, v] of Object.entries(g)) {
      if (!mw[m] && mw[m] !== 0) continue; // month outside requested range
      if (m < patchMonth) pre.push(v); else if (m > patchMonth) post.push(v);
    }
    const mean = a => a.reduce((s, v) => s + v, 0) / a.length;
    const monthsSeen = new Set();
    for (const r of Object.values(byrank))
      for (const m of Object.keys(r)) if (m in mw) monthsSeen.add(m);
    rows.push({
      opp,
      t40: tier['40'], t41: tier['41'], t42: tier['42'],
      comb: num / den,
      spread: Math.max(...present) - Math.min(...present),
      dpatch: pre.length && post.length ? mean(post) - mean(pre) : null,
      nmonths: monthsSeen.size,
    });
  }
  rows.sort((a, b) => a.comb - b.comb);
  return rows;
}

/* mirrors recommend.py main loop; rows sorted by cover descending */
function subTable(idx, char, mw, exclude, tierWeights, oppWeights) {
  const mainRow = combinedRow(idx, char, mw, exclude, tierWeights);
  const worst3 = Object.keys(mainRow).sort((a, b) => mainRow[a] - mainRow[b]).slice(0, 3);
  const results = [];
  for (const sub of Object.keys(idx)) {
    if (sub === char || exclude.has(sub)) continue;
    const row = combinedRow(idx, sub, mw, exclude, tierWeights);
    if (!Object.keys(row).length) continue;
    const perTier = {};
    for (const r of Object.keys(tierWeights))
      perTier[r] = coverage(mainRow, combinedRow(idx, sub, mw, exclude, tierWeights, [r]), oppWeights);
    const w3 = worst3.filter(o => o in row).map(o => row[o]);
    results.push({
      sub,
      cover: coverage(mainRow, row, oppWeights),
      c40: perTier['40'], c41: perTier['41'], c42: perTier['42'],
      spec: specialization(mainRow, row, oppWeights),
      strength: strength(row),
      w3win: w3.length ? w3.reduce((s, v) => s + v, 0) / w3.length * 10 : null,
      corr: correlation(mainRow, row),
      shared: sharedWeaknesses(mainRow, row).length,
    });
  }
  results.sort((a, b) => b.cover - a.cover);
  return { worst3, mainRow, results };
}

if (typeof module !== 'undefined') {
  module.exports = {
    PATCH_MONTH, DEFAULT_TIER_WEIGHTS, RANK_NAMES, TARGET_INJECT,
    monthWeights, wavg, coverage, specialization, strength, correlation, sharedWeaknesses,
    buildIndex, availableMonths, combinedRow, charTable, subTable,
  };
}
